### PR TITLE
chore: continue unmodified CDP Fetch responses instead of fulfilling

### LINKED ISSUE #34467: chore: continue unmodified CDP Fetch responses instead of fulfilling
Split from #34352. Part of proxy-disabled / HTTP/2 (#33846).

## Symptom

Under `CYPRESS_INTERNAL_DISABLE_PROXY=1`, the CDP Fetch transport effectively `Fetch.fulfillRequest`s every response because the synthetic codec always attaches a body and the terminal decision is roughly `fulfilled: response.body !== undefined`. Replacing Chromium's native response handling breaks native browser behavior and enlarges the HTML rewrite failure surface.

## Corrected premise

- Reading the body does **not** force fulfillment — `Fetch.getResponseBody` then `Fetch.continueResponse` is valid (capture spike `gated` mode).
- This issue does **not** stop materializing bodies. It only changes the final transport step: fulfill **iff** the body was modified.

## Goals

1. Track an explicit `bodyModified` (or equivalent) signal; fulfill only when set.
2. Unmodified responses → `Fetch.continueResponse`.
3. Header-only / status-only changes use `continueResponse`'s `responseCode` / `responseHeaders` rather than forcing fulfill.
4. Matched-but-unmutated `cy.intercept` (spy / alias / read-only `req.continue`) continues natively.
5. Test Replay middleware / `notifyResponseStreamReceived` tee untouched.

## Non-goals

- Native download system-test greening (separate issue; run the download gating experiment there).
- Test Replay end-to-end capture verification (separate validation issue).
- Crash cluster / `Invalid InterceptionId` (separate; materialization stays).
- Skipping body reads for SSE / large media (follow-on exclusion list).
- Public API / `res.body` changes.
- Performance work.

## Acceptance criteria

- [ ] Unmodified → `continueResponse`; modified → `fulfillRequest` with correct body/headers.
- [ ] Header-only / status-only changes ride continue params.
- [ ] Spy / alias / read-only continue does not fulfill.
- [ ] Unit coverage on the transport decision.
- [ ] Proxy-enabled path remains green.
- [ ] Crash cluster and downloads AC **not** claimed here.

## Test plan

```bash
yarn workspace @packages/server test-unit -- test/unit/browsers/cdp-fetch-transport_spec.ts
scripts/validate-proxy-middleware-transport.sh both
```

## Related

- Parent tracker: #34352
- Prior closed attempts: #34371, #34408
- Does not replace: #34391 (`strategy:file` fixture downloads)

### LINKED ISSUE #34468: chore: native downloads when the MITM proxy is disabled (CDP Fetch)
Split from #34352. Part of proxy-disabled / HTTP/2 (#33846).

## Symptom

Browser-native downloads (`Page.downloadWillBegin` / Electron `will-download`) fail or never fire under `CYPRESS_INTERNAL_DISABLE_PROXY=1` when responses are fulfilled instead of continued. Fixture/`strategy:file` downloads were addressed separately in #34391 — this issue is **native** download handling only.

## Gate (do before implementing)

Nobody has verified downloads under `getResponseBody` + `continueResponse`. Run a small experiment:

1. Does `Fetch.getResponseBody` succeed at a download's response pause?
2. Does `Fetch.continueResponse` after reading preserve native download handling?

- If **yes** → depend on the continue-unmodified transport fix and assert system downloads green.
- If **no** → document exclusion / alternate path; do not force the continue issue to claim downloads.

## Goals

1. `downloads_spec` (native cases) green under proxy-disabled for Chrome and Electron.
2. Clear written result of the gating experiment on this issue.

## Non-goals

- `strategy:file` / fixture file-server path (#34391).
- Extra-target basic-auth middleware wiring (separate).
- Test Replay capture recovery for download bodies.
- Changing always-buffer policy for all resource types.

## Acceptance criteria

- [ ] Gating experiment results posted on this issue.
- [ ] Native download system-tests green proxy-off (Chrome + Electron), **or** explicit exclusion design accepted and filed as follow-on.
- [ ] Proxy-enabled downloads still green.

## Test plan

```bash
cd system-tests && CYPRESS_INTERNAL_DISABLE_PROXY=1 \
  PATH="$(pwd)/../node_modules/.bin:$PATH" \
  node ./scripts/run.js --glob-in-dir="test" downloads_spec --browser chrome
# repeat electron
```

## Related

- Parent tracker: #34352
- Depends on / coordinates with: continue unmodified CDP responses
- Already done: #34391

### PR DESCRIPTION
**Pull request by cacieprins**
- Closes #34467
- Closes #34468
- Parent tracker: #34352

### Additional details

Under `CYPRESS_INTERNAL_DISABLE_PROXY=1`, the synthetic proxy codec always materializes a response body for the intercept pipeline, so the CDP Fetch codec treated every successful response as fulfilled (`response.body !== undefined` → `Fetch.fulfillRequest`). That replaces Chromium's native response handling and enlarges the HTML rewrite / native-download failure surface.

This change keeps body materialization (Test Replay capture, intercepts, injection still work). It only changes the terminal step:

1. At `Fetch.getResponseBody` time, hash the origin bytes (`length` + `sha256`) into `originalBodyDigest` and **do not** retain the buffer on the pause — peak memory stays at one body copy while the middleware stream drains.
2. After the middleware onion, fulfill **iff** the pipeline's final body digests differently from the origin (`isOriginBody` in [`body-digest.ts`](packages/server/lib/browsers/cdp-protocol/body-digest.ts)). Otherwise `Fetch.continueResponse`.
3. Header-only / status-only edits ride `continueResponse` with middleware headers plus the pause's wire `content-encoding` / `content-length` / `transfer-encoding`.

Matched-but-unmutated `cy.intercept` (spy / alias / read-only `req.continue`) continues natively because the bytes round-trip. Request-stage short-circuits (blocked hosts, internal routes, `strategy:file`, static stubs) still always fulfill. Missing digest fails open to fulfill (today's behavior).

**#34468 gating experiment (reported, not claimed):** `downloads_spec` under `CYPRESS_INTERNAL_DISABLE_PROXY=1` passes for **Chrome and Electron** after this change — native downloads survive `getResponseBody` + `continueResponse`.

**CI promotion (`*-cdp-remediated`):**
- System (chrome + electron): `proxying_spec`, `images_spec`, `form_submissions_spec`, `page_loading_spec`, `interception_spec`, `cache_spec` (plus prior `network_error_handling` / `downloads`)
- Driver (chrome): `security.cy.js`, `redirects.cy.js`, `snapshot.cy.js`, `issues/3890.cy.js`
- Not promoted (still red / flaky under CDP): `cookies_spec`, `non_proxied_spec`, `proxy_correlation_spec` (Chrome CI flake: CorrelateBrowserPreRequest timeouts under concurrent 1MB bodies), driver `encoding` / `csp_headers` / `proxy-logging` / `service-worker`

### Steps to test

```bash
yarn workspace @packages/server test-unit -- cdp-fetch-transport_spec.ts
yarn workspace @packages/server test-unit -- network-runtime_spec.ts
scripts/validate-proxy-middleware-transport.sh both   # driver portion needs free :3500
CYPRESS_INTERNAL_DISABLE_PROXY=1 yarn workspace @tooling/system-tests test:ci test/downloads_spec.ts --browser chrome
```

### How has the user experience changed?

No change for the default MITM proxy path. With the internal `CYPRESS_INTERNAL_DISABLE_PROXY=1` flag, unmodified responses (including spy/alias intercepts) are released natively via CDP `continueResponse` instead of being re-fulfilled from Node.

### PR Tasks

- [x] Is there an associated issue with maintainer approval for PR submission? (#34467 / #34352 HTTP/2 work)
- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

<div><a href="https://cursor.com/agents/bc-5ff97136-483b-4b65-8de9-d9dd90ce4710?cursor_ref=pr_footer&cursor_cta=open_in_web"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-5ff97136-483b-4b65-8de9-d9dd90ce4710&cursor_ref=pr_footer&cursor_cta=open_in_cursor"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

File: packages/server/lib/browsers/cdp-protocol/cdp-fetch-codec.ts [hunk 1]
```diff
@@ -87,8 +88,42 @@ function stripWireEncodingHeaders (headers?: HttpHeaders): HttpHeaders | undefin
...
+
+  return pauseHeaders ? stripHeaderEntries(pauseHeaders, WIRE_ENCODING_HEADERS) : undefined
+}
+
+/**
+ * The browser replays the origin's own wire body on continueResponse, so the
+ * pause's wire encoding headers are the ones that still describe it. Carry them
+ * over the middleware view, which only ever saw the decoded body.
+ */
+function toContinuedHeaders (headers?: HttpHeaders, pauseHeaders?: Protocol.Fetch.HeaderEntry[]): Protocol.Fetch.HeaderEntry[] | undefined {
```
Comment (AtofStryker) [inline packages/server/lib/browsers/cdp-protocol/cdp-fetch-codec.ts:141]: **Omit `responseHeaders` from `continueResponse` when middleware didn't change them.** The request stage already does exactly this (`continueRequestHeaders` diffs via `headersChanged` and omits when clean), but `toContinuedHeaders` sends a full replacement header set on essentially every response — a lowercased, re-folded reconstruction of the origin headers (the casing change is already visible in the updated test expectations). "Don't override what you didn't touch" is strictly less risk and makes the two stages symmetric.

File: packages/server/lib/browsers/cdp-protocol/cdp-fetch-codec.ts [hunk 2]
```diff
@@ -87,8 +88,42 @@ function stripWireEncodingHeaders (headers?: HttpHeaders): HttpHeaders | undefin
...
+  const middlewareHeaders = toResponseHeaders(headers)
+
+  if (!middlewareHeaders) {
```
Comment (AtofStryker) [inline packages/server/lib/browsers/cdp-protocol/cdp-fetch-codec.ts:144]: **Pin the `headers = {}` semantics in `toContinuedHeaders`/`toFulfilledHeaders`.** `toResponseHeaders({})` returns `[]` (truthy), so `res.headers = {}` replaces the origin headers with only the pause's wire headers. I couldn't reach this unintentionally through the real pipeline, but a later refactor treating `[]` as "absent" would silently flip "delete all headers" into "headers untouched." Add a test pinning the intended semantic — or normalize empty→undefined at the decode seam so the edge can't surprise either way.

File: packages/server/lib/browsers/cdp-protocol/cdp-fetch-transport.ts [hunk 3]
```diff
@@ -454,18 +458,16 @@ export class CdpFetchTransport {
     ]
   }
 
-  private fetchResponseBody = async (event: Protocol.Fetch.RequestPausedEvent, sessionId?: string): Promise<Readable> => {
+  private fetchResponseBody = async (event: Protocol.Fetch.RequestPausedEvent, sessionId?: string): Promise<Buffer> => {
     if (isRedirectPause(event)) {
-      return Readable.from([])
+      return Buffer.alloc(0)
```
Comment (AtofStryker) [inline packages/server/lib/browsers/cdp-protocol/cdp-fetch-transport.ts:525]: **Extend the `isRedirectPause` doc by one sentence.** This `Buffer.alloc(0)` now digests as the origin body, and that empty digest is what routes an untouched 3xx to `continueResponse` (and a middleware that writes a body onto a 3xx back to fulfill). The existing comment covers why the body is empty, not that its digest now decides fulfill-vs-continue.

File: packages/server/lib/browsers/cdp-protocol/body-digest.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { createHash } from 'crypto'
+
+export type BodyDigest = { length: number, sha256: string }
+
+export const digestBody = (body: Buffer): BodyDigest => ({
```
Comment (AtofStryker) [inline packages/server/lib/browsers/cdp-protocol/body-digest.ts:8]: **Non-blocking: a short comment on the digest cost.** Worth noting here that hashing runs per response body so nobody re-derives whether it matters. It doesn't — bodies large enough to register are extremely rare, and even then the drag is negligible next to what the CDP path already pays per byte. Suggestion:

```suggestion
// Runs once per response body (twice when lengths match, behind the length
// short-circuit). sha256 sustains ~GB/s, so even rare multi-MB bodies add
// negligible drag next to the base64 decode the CDP transit already costs.
export const digestBody = (body: Buffer): BodyDigest => ({
```

**Review by AtofStryker — COMMENTED** [review]
The core change looks right to me and the test matrix is strong — verbatim continue, mutated fulfill, same-length-different-bytes, header-only, status-only, missing-digest fail-open, and the legacy-pipeline asset round-trip in network-runtime_spec (that last one is the load-bearing assertion that the pipeline actually preserves bytes). Fail-safe direction is correct everywhere: anything unprovable falls back to fulfill, so a rewrite can never be silently dropped.

I left four code-level requests as inline comments. One more thing that isn't diff-scoped:

**Downloads / #34468.** Verified two ways that fulfilling does not break native downloads:

- Raw CDP repro (Chrome 150): a mutated response-stage fulfill downloads the mutated bytes, a request-stage stub downloads, and continue preserves the origin bytes.
- Driver-level under proxy-disabled: `cy.intercept` + `req.continue` mutating `res.body` → the native download completes containing the mutated body (snippet below; I'll attach screenshots in a follow-up comment).

```ts
  it.only('downloads from anchor tag with download attribute', () => {
    cy.intercept('GET', '/fixtures/downloads_records.csv', (req) => {
    // proxy off, mutate the response body and make sure native download works
      req.continue((res) => {
        res.body = 'test'
      })
    })

    cy.exec(`rm -f ${Cypress.config('downloadsFolder')}/downloads_records.csv`)
    cy.readFile(`${Cypress.config('downloadsFolder')}/downloads_records.csv`).should('not.exist')

    // trigger download
    cy.get('[data-cy=download-csv]').click()
    cy.readFile(`${Cypress.config('downloadsFolder')}/downloads_records.csv`)
    // fails: is now 'test'
    .should('contain', '"Joe","Smith"')
  })
```
Screenshot 2026-08-05 at 1 27 58 PM

Screenshot 2026-08-05 at 1 48 36 PM

So there's no intercept-vs-download drift, and I think this PR satisfies #34468's gate — going to post these results there and we should add "closes #34468". 

One carve-out to note when closing:

- Driver `downloads.cy.ts`'s "basic auth protected file that opens in a new tab" still fails under proxy-disabled (the auth prompt opens in the new tab). That's the extra-target basic-auth wiring already listed as a non-goal on #34468, so it needs its own issue and blocks promoting driver downloads.cy.ts to the remediated job until it lands. Worth noting it's basic auth specifically — narrow enough that we could conditionally skip that case under proxy-disabled and leave it out of the MVP entirely. Team decision.

**Comment by AtofStryker** [comment]
@cacieprins I went ahead and added #34468 to issues this PR closes

**Review by AtofStryker — APPROVED** [review]
Approving — all four of my review asks landed, each pinned by tests:

- `continueResponse` now omits `responseHeaders` when middleware didn't meaningfully change them (case/order-insensitive diff), so pass-through responses keep the browser's original headers — including the set-cookie case, which stays fully native now. Strictly better fidelity than what I asked for.
- The `headers = {}` "delete all headers" semantic is pinned on both terminal paths, with the truthiness footgun called out in the test comments.
- The `isRedirectPause` doc now covers that the empty buffer is the origin digest source, which is what routes an untouched 3xx to `continueResponse`.
- Digest-cost comment applied, plus the dedicated body-digest_spec is a nice bonus.

Also appreciate the `record-group-suffix: '-remediated'` catch on the new driver job.

On CI: the `server-performance-tests` failures are all "With Cypress proxy" MITM timing thresholds against the external benchmark host — code this PR doesn't execute — and the percy diffs came in with the develop merge (no frontend changes here). Neither is attributable to this change; everything proxy-disabled is green, including the remediated jobs on the promoted specs.

Downloads/#34468 experiment results are posted on that issue (https://github.com/cypress-io/cypress/issues/34468#issuecomment-5195306911) — with this PR the gate is satisfied, so it can close whenever the team is ready, with the basic-auth new-tab carve-out tracked separately.

**Comment by cacieprins** [comment]
Failing CI is known flake - Merging
